### PR TITLE
Restructure meeting scheduling Algorithm along with Tests

### PR DESCRIPTION
**Pull request by hananeikhelef**
_**Test Driven Development**_
- Restructured the meeting algorithm (break the code into functions )

-  Added some tests 

-  Handle the optional attendees

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/MeetingRequest.java [hunk 1]
```diff
@@ -57,6 +57,15 @@ public void addOptionalAttendee(String attendee) {
       optional_attendees.add(attendee);
     }
   }
+   /**
+   * Returns the duration of the meeting in minutes.
+   */
```
Comment (mctavish) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/MeetingRequest.java]: This comment isn't for this function.

File: walkthroughs/week-5-tdd/project/src/main/webapp/script.js [hunk 2]
```diff
@@ -102,6 +102,7 @@ class TimeRange {
   constructor(start, duration) {
     this.start = start;
     this.duration = duration;
+    this.optional_attendees = [];
```
Comment (mctavish) [inline walkthroughs/week-5-tdd/project/src/main/webapp/script.js]: Why is this needed?

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 3]
```diff
@@ -268,6 +271,63 @@ public void notEnoughRoom() {
     Collection<TimeRange> actual = query.query(events, request);
     Collection<TimeRange> expected = Arrays.asList();
 
+    Assert.assertEquals(expected, actual);
+  }
+  @Test
```
Comment (mctavish) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: Nit: please leave an empty line between } and the Test annotation.  It makes it easier to read.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 4]
```diff
@@ -268,6 +271,63 @@ public void notEnoughRoom() {
...
+    MeetingRequest request =
+        new MeetingRequest(Arrays.asList(PERSON_B,PERSON_A), DURATION_30_MINUTES);
+    request.addOptionalAttendee(PERSON_C);
+
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected =
+        Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false),
+            TimeRange.fromStartEnd(TIME_0930AM, TimeRange.END_OF_DAY, true));
+
     Assert.assertEquals(expected, actual);
```
Comment (mctavish) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:332]: There are definitely more tests that you can come up with here.  What about all attendees being optional?  Other arrangements of the time ranges? Especially try to come up with tests that try to break your code and that verify the edge cases.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -13,61 +13,75 @@
...
-       // whenever no schedulued events exists, we can schedule in any slot  
-        if (eventTimeRanges.size() == 0){
-            timeSlots.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY,TimeRange.END_OF_DAY,true));
-            return timeSlots;
-        }
-        else {
-            for (TimeRange event: eventTimeRanges) {
-                endTime = event.start();
-                if (endTime-startTime >= request.getDuration()) {
+  public Collection<TimeRange> getMeetingRanges(List<TimeRange> eventTimeRange, long duration){
```
Comment (mctavish) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:76]: Go with List<> if the output has an ordering to it (sorted according to start time, in this case).

Comment (mctavish) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:76]: This function could do any of several things based on the signature.  I'd like to see javadoc to make it clear what it does.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -13,61 +13,75 @@
...
+
+        Collection<TimeRange> timeSlots = new ArrayList();
+        int startTime = TimeRange.getTimeInMinutes(0, 0);
+        int endTime = TimeRange.getTimeInMinutes(0, 0);
+
+        for (TimeRange event: eventTimeRange) {
+            endTime = event.start();
+                if (endTime-startTime >= duration) {
```
Comment (mctavish) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Nit: Your indenting is off.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -13,61 +13,75 @@
...
+  }
+    // The algorithm adds timeranges as we go through the events starting from the start day
+  public List<TimeRange> getEventTimeRanges(Collection<Event> events, Collection<String> request){
+              // sort all the time ranges and add  to the list the required attendees in the meeting request
         List<TimeRange> eventTimeRanges = 
                 events.stream().filter(event -> event.getAttendees().stream()
-                .anyMatch(attendee -> request.getAttendees().contains(attendee)))
+                .anyMatch(attendee -> request.contains(attendee)))
                 .map(i -> i.getWhen())
                 .collect(Collectors.toList());
```
Comment (mctavish) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: The way this is formatted makes it very hard for me to read this.

The following is an example of how this could be formatted to make it more obvious as to which functions are being applied to which collections:
```
        events.stream()
            .filter(event -> event.getAttendees().stream()
                .anyMatch(attendee -> request.contains(attendee)))
            .map(i -> i.getWhen())
            .collect(Collectors.toList());
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -13,61 +13,75 @@
...
         if(request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
             return timeSlots;
         }
 
-        // sort all the time ranges and add  to the list the required attendees in the meeting request
+        List<TimeRange> eventTimeRanges = getEventTimeRanges(events,request.getAllAttendees());
+        Collection<TimeRange> meetingRanges = getMeetingRanges(eventTimeRanges,request.getDuration());
+
+       // whenever no schedulued events exists, we can schedule in any slot  
+        if (meetingRanges.isEmpty() && !request.getAllAttendees().isEmpty()){
```
Comment (mctavish) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Strange conditional.  When would you be hitting this codepoint if getAllAttendees() is empty?  That sounds like a special case to be handled above.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 9]
```diff
@@ -13,61 +13,75 @@
...
         // check for invalid duration
         if(request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
             return timeSlots;
         }
 
-        // sort all the time ranges and add  to the list the required attendees in the meeting request
+        List<TimeRange> eventTimeRanges = getEventTimeRanges(events,request.getAllAttendees());
+        Collection<TimeRange> meetingRanges = getMeetingRanges(eventTimeRanges,request.getDuration());
+
+       // whenever no schedulued events exists, we can schedule in any slot
```
Comment (mctavish) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: This comment confuses me.  Do you mean that if there were no eligible meeting slots, try again without the optional attendees?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 10]
```diff
@@ -13,61 +13,75 @@
...
+        if (meetingRanges.isEmpty() && !request.getAllAttendees().isEmpty()){
+            List<TimeRange> blockedTimeRanges = 
+                getEventTimeRanges(events, request.getAttendees());
+            Collection<TimeRange> meetingTime = 
+                getMeetingRanges(blockedTimeRanges, request.getDuration());
+            return meetingTime;
+        }
+        else {
+            return meetingRanges;
+        }
```
Comment (mctavish) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:55]: To make it easier to read your code, prefer to invert the conditional here.  Instead of this:

```
if (condition) {
  // do a whole bunch of stuff
  return
} else {
  return
}
```

Prefer this:
```
if ( not condition ) {
  return
}

// do that whole bunch of stuff
return
```

Alternatively, if you can swing it, this is even better:
```
if (condition) {
  // do a simple version of stuff
}
return
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 11]
```diff
@@ -13,61 +13,75 @@
...
-        // sort all the time ranges and add  to the list the required attendees in the meeting request
+        List<TimeRange> eventTimeRanges = getEventTimeRanges(events,request.getAllAttendees());
+        Collection<TimeRange> meetingRanges = getMeetingRanges(eventTimeRanges,request.getDuration());
+
+       // whenever no schedulued events exists, we can schedule in any slot  
+        if (meetingRanges.isEmpty() && !request.getAllAttendees().isEmpty()){
+            List<TimeRange> blockedTimeRanges = 
+                getEventTimeRanges(events, request.getAttendees());
+            Collection<TimeRange> meetingTime = 
+                getMeetingRanges(blockedTimeRanges, request.getDuration());
```
Comment (mctavish) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:50]: No need to declare new variables here.  You can just clobber eventTimeRanges and meetingRanges from earlier, which will make this block of code shorter and save you the need to come up with new variable names.

**Review by mctavish — COMMENTED** [review]
Looks good!  I've left a few comments here, mostly to make the code easier to read (and thus maintain if this were part of a long-lasting project) and asking for more thorough test coverage.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/MeetingRequest.java (hunk 1 above)
Comment (hananeikhelef) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/MeetingRequest.java]: Changes have been made :)

File: walkthroughs/week-5-tdd/project/src/main/webapp/script.js (hunk 2 above)
Comment (hananeikhelef) [reply walkthroughs/week-5-tdd/project/src/main/webapp/script.js]: My bad, i added and forgot to delete it after realizing there is already an array for optional attendees in **MeetingRequest.java** changes have been made.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 3 above)
Comment (hananeikhelef) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: Done.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 7 above)
Comment (hananeikhelef) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Thank you  for the insight, changes have been made.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 4 above)
Comment (hananeikhelef) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:332]: I added more tests to cover more edge cases :)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 8 above)
Comment (hananeikhelef) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I added this condition to make it pass an existing test  that wasn't passing.
